File: scripts/validate_inputs.py
#!/usr/bin/env python3
"""Validate input CSV files for the 2026 World Cup prediction report.

The validator enforces the two fixed execution parameters:
1. World Cup schedule data must use the real schedule schema.
2. Historical training data must be real match-result data cut off at 2026-06-10.
"""

from __future__ import annotations

import argparse
import csv
from datetime import date, datetime
from pathlib import Path
from typing import Iterable


SCHEDULE_REQUIRED = [
    "match_id",
    "stage",
    "group",
    "match_date",
    "team_1",
    "team_2",
    "venue",
    "host_country",
    "team_1_slot",
    "team_2_slot",
]

HISTORY_REQUIRED = [
    "date",
    "home_team",
    "away_team",
    "home_score",
    "away_score",
    "tournament",
    "country",
    "neutral",
]

GROUPS_REQUIRED = ["group", "slot", "team"]

CUTOFF = date(2026, 6, 10)
TRAIN_START = date(2010, 1, 1)
TRAIN_END = date(2024, 12, 31)
TEST_START = date(2025, 1, 1)
TEST_END = CUTOFF


def parse_date(value: str, field: str, row_number: int) -> date:
    try:
        return datetime.strptime(value.strip(), "%Y-%m-%d").date()
    except ValueError as exc:
        raise ValueError(f"row {row_number}: invalid {field} '{value}', expected YYYY-MM-DD") from exc


def read_rows(path: Path) -> tuple[list[str], list[dict[str, str]]]:
    with path.open("r", encoding="utf-8-sig", newline="") as handle:
        reader = csv.DictReader(handle)
        if reader.fieldnames is None:
            raise ValueError(f"{path} has no header row")
        rows = list(reader)
        return list(reader.fieldnames), rows


def require_columns(path: Path, columns: Iterable[str], required: list[str]) -> None:
    missing = [column for column in required if column not in columns]
    if missing:
        raise ValueError(f"{path} is missing required columns: {', '.join(missing)}")


def validate_schedule(path: Path) -> None:
    columns, rows = read_rows(path)
    require_columns(path, columns, SCHEDULE_REQUIRED)
    if not rows:
        raise ValueError(f"{path} contains no schedule rows")

    stages = set()
    groups = set()
    for idx, row in enumerate(rows, start=2):
        match_date = parse_date(row["match_date"], "match_date", idx)
        if match_date < date(2026, 1, 1):
            raise ValueError(f"row {idx}: match_date predates 2026")
        stages.add(row["stage"].strip())
        if row["group"].strip():
            groups.add(row["group"].strip())
        for field in ["match_id", "stage", "team_1", "team_2", "venue", "host_country"]:
            if not row[field].strip():
                raise ValueError(f"row {idx}: {field} must not be empty")

    if "group" not in stages:
        raise ValueError("schedule must include group-stage rows with stage='group'")
    expected_groups = set("ABCDEFGHIJKL")
    if groups and not groups.issubset(expected_groups):
        invalid = ", ".join(sorted(groups - expected_groups))
        raise ValueError(f"schedule contains invalid group labels: {invalid}")


def validate_groups(path: Path) -> None:
    columns, rows = read_rows(path)
    require_columns(path, columns, GROUPS_REQUIRED)
    if len(rows) != 48:
        raise ValueError(f"{path} must contain exactly 48 teams, found {len(rows)}")

    expected_groups = set("ABCDEFGHIJKL")
    group_counts: dict[str, int] = {}
    teams: set[str] = set()
    for idx, row in enumerate(rows, start=2):
        group = row["group"].strip()
        slot = row["slot"].strip()
        team = row["team"].strip()
        if group not in expected_groups:
            raise ValueError(f"row {idx}: invalid group '{group}'")
        if not slot.startswith(group):
            raise ValueError(f"row {idx}: slot '{slot}' must start with group '{group}'")
        if not team:
            raise ValueError(f"row {idx}: team must not be empty")
        if team in teams:
            raise ValueError(f"row {idx}: duplicate team '{team}'")
        teams.add(team)
        group_counts[group] = group_counts.get(group, 0) + 1

    missing_groups = expected_groups - set(group_counts)
    if missing_groups:
        raise ValueError(f"missing groups: {', '.join(sorted(missing_groups))}")
    bad_groups = {group: count for group, count in group_counts.items() if count != 4}
    if bad_groups:
        details = ", ".join(f"{group}={count}" for group, count in sorted(bad_groups.items()))
        raise ValueError(f"each group must contain 4 teams; invalid counts: {details}")


def validate_history(path: Path) -> None:
    columns, rows = read_rows(path)
    require_columns(path, columns, HISTORY_REQUIRED)
    if not rows:
        raise ValueError(f"{path} contains no historical match rows")

    train_count = 0
    test_count = 0
    for idx, row in enumerate(rows, start=2):
        match_date = parse_date(row["date"], "date", idx)
        if match_date > CUTOFF:
            raise ValueError(f"row {idx}: match date {match_date} is after cutoff {CUTOFF}")
        if TRAIN_START <= match_date <= TRAIN_END:
            train_count += 1
        if TEST_START <= match_date <= TEST_END:
            test_count += 1
        for field in ["home_team", "away_team", "tournament", "country"]:
            if not row[field].strip():
                raise ValueError(f"row {idx}: {field} must not be empty")
        for field in ["home_score", "away_score"]:
            try:
                score = int(row[field])
            except ValueError as exc:
                raise ValueError(f"row {idx}: {field} must be an integer") from exc
            if score < 0:
                raise ValueError(f"row {idx}: {field} must be non-negative")
        neutral = row["neutral"].strip().lower()
        if neutral not in {"true", "false", "1", "0", "yes", "no"}:
            raise ValueError(f"row {idx}: neutral must be boolean-like")

    if train_count == 0:
        raise ValueError("history data has no training rows in 2010-01-01..2024-12-31")
    if test_count == 0:
        raise ValueError("history data has no test rows in 2025-01-01..2026-06-10")


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--groups", type=Path, default=Path("data/worldcup_2026_groups.csv"))
    parser.add_argument("--schedule", type=Path, default=Path("data/worldcup_2026_schedule.csv"))
    parser.add_argument("--history", type=Path, default=Path("data/historical_matches.csv"))
    args = parser.parse_args()

    validate_groups(args.groups)
    validate_schedule(args.schedule)
    validate_history(args.history)
    print("Input validation passed.")
    print(f"Groups: {args.groups}")
    print(f"Schedule: {args.schedule}")
    print(f"History: {args.history}")
    print(f"Date cutoff: {CUTOFF.isoformat()}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: R/validate_inputs.R
#!/usr/bin/env Rscript

# Validate input CSV files for the 2026 World Cup prediction report.
# This script uses base R only.

args <- commandArgs(trailingOnly = TRUE)

get_arg <- function(flag, default) {
  pos <- match(flag, args)
  if (!is.na(pos) && pos < length(args)) {
    return(args[[pos + 1]])
  }
  default
}

groups_path <- get_arg("--groups", "data/worldcup_2026_groups.csv")
schedule_path <- get_arg("--schedule", "data/worldcup_2026_schedule.csv")
history_path <- get_arg("--history", "data/historical_matches.csv")
results_path <- get_arg("--results", "data/worldcup_2026_results_asof_2026-06-20.csv")
annex_path <- get_arg("--annex-c", "data/annex_c_full_mapping.csv")
allow_unconfirmed_data <- "--allow-unconfirmed-data" %in% args

required_groups <- c("group", "slot", "team")
required_schedule <- c(
  "match_id", "stage", "group", "match_date", "team_1", "team_2",
  "venue", "host_country", "team_1_slot", "team_2_slot"
)
required_history <- c(
  "date", "home_team", "away_team", "home_score", "away_score",
  "tournament", "country", "neutral"
)

cutoff <- as.Date("2026-06-10")
train_start <- as.Date("2010-01-01")
train_end <- as.Date("2024-12-31")
test_start <- as.Date("2025-01-01")
test_end <- cutoff

stop_missing <- function(path) {
  if (!file.exists(path)) {
    stop(sprintf("Required file not found: %s", path), call. = FALSE)
  }
}

require_columns <- function(df, required, path) {
  missing <- setdiff(required, names(df))
  if (length(missing) > 0) {
    stop(sprintf("%s is missing required columns: %s", path, paste(missing, collapse = ", ")), call. = FALSE)
  }
}

read_required_csv <- function(path) {
  stop_missing(path)
  read.csv(path, stringsAsFactors = FALSE, fileEncoding = "UTF-8-BOM")
}

validate_groups <- function(path) {
  df <- read_required_csv(path)
  require_columns(df, required_groups, path)
  if (nrow(df) != 48) {
    stop(sprintf("%s must contain exactly 48 teams, found %s", path, nrow(df)), call. = FALSE)
  }
  expected_groups <- LETTERS[1:12]
  bad_groups <- setdiff(unique(df$group), expected_groups)
  if (length(bad_groups) > 0) {
    stop(sprintf("Invalid group labels: %s", paste(bad_groups, collapse = ", ")), call. = FALSE)
  }
  counts <- table(df$group)
  if (length(counts) != 12 || any(counts != 4)) {
    stop("Groups file must contain 12 groups and exactly 4 teams per group.", call. = FALSE)
  }
  if (any(duplicated(df$team))) {
    stop("Groups file contains duplicate teams.", call. = FALSE)
  }
  slot_ok <- mapply(function(g, s) startsWith(s, g), df$group, df$slot)
  if (!all(slot_ok)) {
    stop("Every slot must start with its group letter, e.g. A1 for group A.", call. = FALSE)
  }
  invisible(TRUE)
}

validate_schedule <- function(path) {
  df <- read_required_csv(path)
  require_columns(df, required_schedule, path)
  if (nrow(df) == 0) {
    stop("Schedule file contains no rows.", call. = FALSE)
  }
  dates <- as.Date(df$match_date)
  if (any(is.na(dates))) {
    stop("Schedule contains invalid match_date values.", call. = FALSE)
  }
  if (any(dates < as.Date("2026-01-01"))) {
    stop("Schedule contains dates before 2026.", call. = FALSE)
  }
  if (!any(tolower(df$stage) == "group")) {
    stop("Schedule must include group-stage rows with stage='group'.", call. = FALSE)
  }
  invalid_groups <- setdiff(unique(df$group[df$group != ""]), LETTERS[1:12])
  if (length(invalid_groups) > 0) {
    stop(sprintf("Schedule contains invalid group labels: %s", paste(invalid_groups, collapse = ", ")), call. = FALSE)
  }
  invisible(TRUE)
}

validate_history <- function(path) {
  df <- read_required_csv(path)
  require_columns(df, required_history, path)
  if (nrow(df) == 0) {
    stop("Historical match file contains no rows.", call. = FALSE)
  }
  dates <- as.Date(df$date)
  if (any(is.na(dates))) {
    stop("Historical match file contains invalid date values.", call. = FALSE)
  }
  if (any(dates > cutoff)) {
    stop(sprintf("Historical match file contains matches after cutoff %s.", cutoff), call. = FALSE)
  }
  if (!any(dates >= train_start & dates <= train_end)) {
    stop("Historical match file has no training rows in 2010-01-01..2024-12-31.", call. = FALSE)
  }
  if (!any(dates >= test_start & dates <= test_end)) {
    stop("Historical match file has no test rows in 2025-01-01..2026-06-10.", call. = FALSE)
  }
  if (any(is.na(as.integer(df$home_score))) || any(is.na(as.integer(df$away_score)))) {
    stop("Scores must be integers.", call. = FALSE)
  }
  if (any(as.integer(df$home_score) < 0) || any(as.integer(df$away_score) < 0)) {
    stop("Scores must be non-negative.", call. = FALSE)
  }
  neutral_values <- tolower(trimws(as.character(df$neutral)))
  allowed <- c("true", "false", "1", "0", "yes", "no")
  if (any(!neutral_values %in% allowed)) {
    stop("neutral must be boolean-like: true/false/1/0/yes/no.", call. = FALSE)
  }
  invisible(TRUE)
}

validate_results <- function(path) {
  df <- read_required_csv(path)
  required <- c("match_date", "group", "team_1", "team_2", "team_1_score", "team_2_score")
  require_columns(df, required, path)
  if (any(is.na(as.Date(df$match_date)))) stop("Results contain invalid match_date values.", call. = FALSE)
  if (any(is.na(as.integer(df$team_1_score))) || any(is.na(as.integer(df$team_2_score)))) {
    stop("Result scores must be integers.", call. = FALSE)
  }
  if ("stage" %in% names(df) && any(tolower(df$stage) != "group")) {
    knockout <- df[tolower(df$stage) != "group", , drop = FALSE]
    require_columns(knockout, c("match_id", "winner", "decided_by"), path)
    if (any(knockout$match_id == "") || any(knockout$winner == "")) {
      stop("Every knockout result must include match_id and winner.", call. = FALSE)
    }
    allowed <- c("90min", "extra_time", "penalties")
    if (any(!knockout$decided_by %in% allowed)) {
      stop("decided_by must be 90min, extra_time, or penalties.", call. = FALSE)
    }
  }
  invisible(TRUE)
}

canonical_groups <- function(x) paste(sort(unique(strsplit(toupper(gsub("[^A-L]", "", x)), "")[[1]])), collapse = "")

validate_annex <- function(path) {
  if (!file.exists(path)) {
    if (allow_unconfirmed_data) {
      warning("Annex C mapping is absent; validation continues only for debug mode.", call. = FALSE)
      return(invisible(FALSE))
    }
    stop(sprintf("Official Annex C mapping not found: %s", path), call. = FALSE)
  }
  df <- read_required_csv(path)
  require_columns(df, c("qualified_groups", "match_id", "third_group"), path)
  df$qualified_groups <- vapply(df$qualified_groups, canonical_groups, character(1))
  if (length(unique(df$qualified_groups)) != choose(12, 8)) {
    stop("Annex C must contain 495 unique combinations.", call. = FALSE)
  }
  if (any(table(df$qualified_groups) != 8)) {
    stop("Every Annex C combination must contain exactly eight assignments.", call. = FALSE)
  }
  invisible(TRUE)
}

validate_groups(groups_path)
validate_schedule(schedule_path)
validate_history(history_path)
validate_results(results_path)
validate_annex(annex_path)

cat("Input validation passed.\n")
cat(sprintf("Groups: %s\n", groups_path))
cat(sprintf("Schedule: %s\n", schedule_path))
cat(sprintf("History: %s\n", history_path))
cat(sprintf("Results: %s\n", results_path))
cat(sprintf("Annex C: %s\n", ifelse(file.exists(annex_path), annex_path, "missing (debug only)")))
cat(sprintf("Date cutoff: %s\n", cutoff))
